hdf4 build fails on aarch64-darwin

**Describe the bug**

Machine type detection at compile time seems inadequate for this system type, see errors below. #108347 added aarch64 support via a patch borrowed from Fedora, unsure what further changes might be necessary.

I wouldn't be particularly bothered, but GDAL needs hdf4, so this blocks building that library and PostGIS etc.

**To Reproduce**
Steps to reproduce the behavior:
1. `nix-build --argstr localSystem aarch64-darwin -A hdf4`

**Expected behavior**
Successful build.

**Additional context**

```
-- Build files have been written to: /tmp/nix-build-hdf-4.2.15.drv-0/hdf-4.2.15/build
cmake: enabled parallel building
building
build flags: -j4 -l4 SHELL=/nix/store/ldfws5lqyz0irmb3bcvnjawyhsva38xq-bash-4.4-p23/bin/bash
Scanning dependencies of target xdr-shared
[  1%] Building C object mfhdf/xdr/CMakeFiles/xdr-shared.dir/xdr.c.o
[  1%] Building C object mfhdf/xdr/CMakeFiles/xdr-shared.dir/xdrfloat.c.o
[  1%] Building C object mfhdf/xdr/CMakeFiles/xdr-shared.dir/xdrstdio.c.o
/tmp/nix-build-hdf-4.2.15.drv-0/hdf-4.2.15/mfhdf/xdr/xdrstdio.c:97:17: warning: assigning to 'struct xdr_ops *' from 'const struct xdr_ops *' discards qualifiers [-Wincompatible-pointer-types-discards-qualifiers]
    xdrs->x_ops = &xdrstdio_ops;
                ^ ~~~~~~~~~~~~~
1 warning generated.
[  1%] Building C object mfhdf/xdr/CMakeFiles/xdr-shared.dir/xdrarray.c.o
[  1%] Linking C shared library ../../bin/libxdr.dylib
[  1%] Built target xdr-shared
Scanning dependencies of target hdf-static
[  1%] Building C object hdf/src/CMakeFiles/hdf-static.dir/atom.c.o
In file included from /tmp/nix-build-hdf-4.2.15.drv-0/hdf-4.2.15/hdf/src/atom.c:59:
In file included from /tmp/nix-build-hdf-4.2.15.drv-0/hdf-4.2.15/hdf/src/hdf.h:20:
/tmp/nix-build-hdf-4.2.15.drv-0/hdf-4.2.15/hdf/src/hdfi.h:1349:1: error: unknown type name 'If'
If you get an error on this line more than one machine type has been defined.
^
/tmp/nix-build-hdf-4.2.15.drv-0/hdf-4.2.15/hdf/src/hdfi.h:1349:7: error: expected ';' after top level declarator
If you get an error on this line more than one machine type has been defined.
      ^
      ;
/tmp/nix-build-hdf-4.2.15.drv-0/hdf-4.2.15/hdf/src/hdfi.h:1352:9: warning: 'GOT_MACHINE' macro redefined [-Wmacro-redefined]
#define GOT_MACHINE
        ^
/tmp/nix-build-hdf-4.2.15.drv-0/hdf-4.2.15/hdf/src/hdfi.h:511:9: note: previous definition is here
#define GOT_MACHINE 1
        ^
/tmp/nix-build-hdf-4.2.15.drv-0/hdf-4.2.15/hdf/src/hdfi.h:1356:9: warning: 'DF_MT' macro redefined [-Wmacro-redefined]
#define DF_MT             DFMT_LINUXAARCH64
        ^
/tmp/nix-build-hdf-4.2.15.drv-0/hdf-4.2.15/hdf/src/hdfi.h:503:9: note: previous definition is here
#define DF_MT DFMT_APPLE_INTEL
        ^
/tmp/nix-build-hdf-4.2.15.drv-0/hdf-4.2.15/hdf/src/hdfi.h:1358:28: warning: redefinition of typedef 'VOIDP' is a C11 feature [-Wtypedef-redefinition]
typedef void              *VOIDP;
                           ^
/tmp/nix-build-hdf-4.2.15.drv-0/hdf-4.2.15/hdf/src/hdfi.h:527:26: note: previous definition is here
typedef void            *VOIDP;
                         ^
/tmp/nix-build-hdf-4.2.15.drv-0/hdf-4.2.15/hdf/src/hdfi.h:1359:28: warning: redefinition of typedef '_fcd' is a C11 feature [-Wtypedef-redefinition]
typedef char              *_fcd;
                           ^
/tmp/nix-build-hdf-4.2.15.drv-0/hdf-4.2.15/hdf/src/hdfi.h:528:26: note: previous definition is here
typedef char            *_fcd;
                         ^
/tmp/nix-build-hdf-4.2.15.drv-0/hdf-4.2.15/hdf/src/hdfi.h:1360:27: warning: redefinition of typedef 'char8' is a C11 feature [-Wtypedef-redefinition]
typedef char              char8;
                          ^
/tmp/nix-build-hdf-4.2.15.drv-0/hdf-4.2.15/hdf/src/hdfi.h:529:25: note: previous definition is here
typedef char            char8;
                        ^
/tmp/nix-build-hdf-4.2.15.drv-0/hdf-4.2.15/hdf/src/hdfi.h:1361:27: warning: redefinition of typedef 'uchar8' is a C11 feature [-Wtypedef-redefinition]
typedef unsigned char     uchar8;
                          ^
/tmp/nix-build-hdf-4.2.15.drv-0/hdf-4.2.15/hdf/src/hdfi.h:530:25: note: previous definition is here
typedef unsigned char   uchar8;
                        ^
/tmp/nix-build-hdf-4.2.15.drv-0/hdf-4.2.15/hdf/src/hdfi.h:1362:27: warning: redefinition of typedef 'int8' is a C11 feature [-Wtypedef-redefinition]
typedef char              int8;
                          ^
/tmp/nix-build-hdf-4.2.15.drv-0/hdf-4.2.15/hdf/src/hdfi.h:531:25: note: previous definition is here
typedef char            int8;
                        ^
/tmp/nix-build-hdf-4.2.15.drv-0/hdf-4.2.15/hdf/src/hdfi.h:1363:27: warning: redefinition of typedef 'uint8' is a C11 feature [-Wtypedef-redefinition]
typedef unsigned char     uint8;
                          ^
/tmp/nix-build-hdf-4.2.15.drv-0/hdf-4.2.15/hdf/src/hdfi.h:532:25: note: previous definition is here
typedef unsigned char   uint8;
                        ^
/tmp/nix-build-hdf-4.2.15.drv-0/hdf-4.2.15/hdf/src/hdfi.h:1364:27: warning: redefinition of typedef 'int16' is a C11 feature [-Wtypedef-redefinition]
typedef short int         int16;
                          ^
/tmp/nix-build-hdf-4.2.15.drv-0/hdf-4.2.15/hdf/src/hdfi.h:533:25: note: previous definition is here
typedef short           int16;
                        ^
/tmp/nix-build-hdf-4.2.15.drv-0/hdf-4.2.15/hdf/src/hdfi.h:1365:28: warning: redefinition of typedef 'uint16' is a C11 feature [-Wtypedef-redefinition]
typedef unsigned short int uint16;
                           ^
/tmp/nix-build-hdf-4.2.15.drv-0/hdf-4.2.15/hdf/src/hdfi.h:534:25: note: previous definition is here
typedef unsigned short  uint16;
                        ^
/tmp/nix-build-hdf-4.2.15.drv-0/hdf-4.2.15/hdf/src/hdfi.h:1366:27: warning: redefinition of typedef 'int32' is a C11 feature [-Wtypedef-redefinition]
typedef int               int32;
                          ^
/tmp/nix-build-hdf-4.2.15.drv-0/hdf-4.2.15/hdf/src/hdfi.h:535:25: note: previous definition is here
typedef int             int32;
                        ^
/tmp/nix-build-hdf-4.2.15.drv-0/hdf-4.2.15/hdf/src/hdfi.h:1367:27: warning: redefinition of typedef 'uint32' is a C11 feature [-Wtypedef-redefinition]
typedef unsigned int      uint32;
                          ^
/tmp/nix-build-hdf-4.2.15.drv-0/hdf-4.2.15/hdf/src/hdfi.h:536:25: note: previous definition is here
typedef unsigned int    uint32;
                        ^
/tmp/nix-build-hdf-4.2.15.drv-0/hdf-4.2.15/hdf/src/hdfi.h:1368:27: warning: redefinition of typedef 'intn' is a C11 feature [-Wtypedef-redefinition]
typedef int               intn;
                          ^
/tmp/nix-build-hdf-4.2.15.drv-0/hdf-4.2.15/hdf/src/hdfi.h:537:25: note: previous definition is here
typedef int             intn;
                        ^
/tmp/nix-build-hdf-4.2.15.drv-0/hdf-4.2.15/hdf/src/hdfi.h:1369:27: warning: redefinition of typedef 'uintn' is a C11 feature [-Wtypedef-redefinition]
typedef unsigned int      uintn;
                          ^
/tmp/nix-build-hdf-4.2.15.drv-0/hdf-4.2.15/hdf/src/hdfi.h:538:25: note: previous definition is here
typedef unsigned int    uintn;
                        ^
/tmp/nix-build-hdf-4.2.15.drv-0/hdf-4.2.15/hdf/src/hdfi.h:1370:27: warning: redefinition of typedef 'intf' is a C11 feature [-Wtypedef-redefinition]
typedef int               intf;     /* size of INTEGERs in Fortran compiler */
                          ^
/tmp/nix-build-hdf-4.2.15.drv-0/hdf-4.2.15/hdf/src/hdfi.h:541:25: note: previous definition is here
typedef int             intf;     /* size of INTEGERs in Fortran compiler */
                        ^
/tmp/nix-build-hdf-4.2.15.drv-0/hdf-4.2.15/hdf/src/hdfi.h:1371:27: warning: redefinition of typedef 'float32' is a C11 feature [-Wtypedef-redefinition]
typedef float             float32;
                          ^
/tmp/nix-build-hdf-4.2.15.drv-0/hdf-4.2.15/hdf/src/hdfi.h:539:25: note: previous definition is here
typedef float           float32;
                        ^
/tmp/nix-build-hdf-4.2.15.drv-0/hdf-4.2.15/hdf/src/hdfi.h:1372:27: warning: redefinition of typedef 'float64' is a C11 feature [-Wtypedef-redefinition]
typedef double            float64;
                          ^
/tmp/nix-build-hdf-4.2.15.drv-0/hdf-4.2.15/hdf/src/hdfi.h:540:25: note: previous definition is here
typedef double          float64;
                        ^
/tmp/nix-build-hdf-4.2.15.drv-0/hdf-4.2.15/hdf/src/hdfi.h:1373:27: warning: redefinition of typedef 'hdf_pint_t' is a C11 feature [-Wtypedef-redefinition]
typedef long              hdf_pint_t;   /* an integer the same size as a pointer */
                          ^
/tmp/nix-build-hdf-4.2.15.drv-0/hdf-4.2.15/hdf/src/hdfi.h:542:25: note: previous definition is here
typedef long            hdf_pint_t;   /* an integer the same size as a pointer */
                        ^
18 warnings and 2 errors generated.
make[2]: *** [hdf/src/CMakeFiles/hdf-static.dir/build.make:82: hdf/src/CMakeFiles/hdf-static.dir/atom.c.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:1394: hdf/src/CMakeFiles/hdf-static.dir/all] Error 2
make: *** [Makefile:182: all] Error 2
builder for '/nix/store/ql6w6yzf3c07kq87a7nyirbhgyn1d356-hdf-4.2.15.drv' failed with exit code 2
error: build of '/nix/store/ql6w6yzf3c07kq87a7nyirbhgyn1d356-hdf-4.2.15.drv' failed
```

**Notify maintainers**

@knedlsepp

**Metadata**
 - system: `"x86_64-darwin"`
 - host os: `Darwin 20.5.0, macOS 10.16`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.3.10`
 - channels(steve): `"darwin, nixpkgs-21.11pre294296.12e7af1cfb3"`
 - nixpkgs: `/Users/steve/.nix-defexpr/channels/nixpkgs`
 - 
Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
  - hdf4
# a list of nixos modules affected by the problem
module:
```

--- a/pkgs/tools/misc/hdf4/darwin-aarch64.patch
+++ b/pkgs/tools/misc/hdf4/darwin-aarch64.patch
@@ -1,0 +1,11 @@
+--- a/hdf/src/hdfi.h	2021-06-16 16:31:31.000000000 +1200
++++ b/hdf/src/hdfi.h	2021-06-16 16:42:26.000000000 +1200
+@@ -1343,7 +1343,7 @@
+ #endif /* IA64 */
+ 
+ /* Linux AArch64 */
+-#if defined __aarch64__
++#if defined __aarch64__ && !defined __APPLE__
+ 
+ #ifdef GOT_MACHINE
+ If you get an error on this line more than one machine type has been defined.

--- a/pkgs/tools/misc/hdf4/default.nix
+++ b/pkgs/tools/misc/hdf4/default.nix
@@ -44,6 +44,7 @@ stdenv.mkDerivation rec {
       url = "https://src.fedoraproject.org/rpms/hdf/raw/edbe5f49646b609f5bc9aeeee5a2be47e9556e8c/f/hdf-aarch64.patch";
       sha256 = "112svcsilk16ybbsi8ywnxfl2p1v44zh3rfn4ijnl8z08vfqrvvs";
     })
+    ./darwin-aarch64.patch
   ];
 
   nativeBuildInputs = [

